pynobo.py

```diff
@@ -352,86 +417,68 @@ def send_command(self, commands):
-        """Get a response string from the hub and reformat string list before returning it
-        
-        Keyword arguments:  
-        keep_alive -- send handshake every ~14 seconds  
-        bufsize -- maximum bytes to receive from the socket
+        :param bufsize: maximum bytes to receive from the socket  (default 4096)
 
-        Returns: a string list(s) of responses
+        :return: a list of string responses
         """
         response = b''
         while response[-1:] != b'\r':
             try:
-                response += self.client.recv(bufsize)
-            except socket.timeout:
-                # handshake needs to be sent every < 30 sec, preferably every 14 seconds
-                if keep_alive:
-                    now = time.time()
-                    if now - self.last_handshake > 14:
-                        self.send_command([self.API.HANDSHAKE])
-                        self.last_handshake = now
-                else:
-                    break
+                response += await self.reader.read(bufsize)
             except ConnectionError as e:
                 self.logger.info('lost connection to hub (%s)', e)
-                self.socket_connected.clear()
+                await self.close()
                 break
 
         # Handle more than one response in one receive
-        response_list = str(response, 'utf-8').split('\r')
+        response_list = str(response.decode("utf-8")).split('\r')
         response = []
         for r in response_list:
             if r:
                 response.append(r.split(' '))
 
         return response
 
-    def socket_receive(self):
-        """The task running in daemon thread"""
-        while not self.socket_receive_exit_flag.is_set():
-            try:
-                if self.socket_connected.is_set():
-                    resp = self.get_response(True)
-                    for r in resp:
-                        self.logger.debug('received: %s', r)
-
-                        if r[0] == self.API.HANDSHAKE:
-                            pass # Handshake, no action needed
-
-                        elif r[0][0] == 'E':
-                            self.logger.error('error! what did you do? %s', r)
-                            #TODO: Raise something here?
-
-                        else:
-                            self.response_handler(r)
-
-                else:
-                    self.reconnect_hub()
-            except Exception as e:
-                # Ops, now we have real problems
-                self.logger.error("Unhandeled exception: %s", e, exc_info=1)
-                # Just disconnect (in stead of risking an infinite reconnect loop)
-                self.socket_receive_exit_flag.set()
-
-        self.logger.info('receive thread exited')
+    async def socket_receive(self):
+        try:
+            while True:
+                responses = await self.get_response()
+                for r in responses:
```

Comment:
Callbacks could be called after this loop, if we ever hit the `self.response_handler(r)` call, should be easy to track with a boolean.

Reply:
Good idea - that will limit callbacks in case of `get_response`

Reply:
Done.

Reply:
Note - I changed `get_response()` to only return one response at a time (by changing the reader method to read until `\r`). This simplified the logic for handling responses.